[TG-1404] Specific implementation of cover basic block for Java

romainbrenguier opened the pull request:
The current implementation of cover_basic_blockt does a first pass on the goto-program without using the bytecode index information, and then later select instruction for bytecode indices.
The process is a bit complicated and sometimes instruments instruction that are after an IF of the same bytecode indices, leading to wrong information about bytecode coverage.
This PR intruduce an alternative cover_basice_block_javat, which for Java use the bytecode information to determine what are the blocks of the program, and selects the first instruction with that bytecode index as a representative.
This strategy is simple and ensures each goal corresponds to a bytecode index coverage.

romainbrenguier committed:
Put interface to cover_blocks in virtual class

This will allow to have two implementations for this functor.

romainbrenguier committed:
Refer to interface rather than cover_basic_blockst

romainbrenguier committed:
cover_block implementation using bytecode location

This is a specific implementation of cover block for Java.
It uses the bytecode index to determine blocks and replace the use of
select_unique_java_bytecode_indices.

tautschnig commented on src/goto-instrument/cover_basic_blocks.cpp:
```diff
@@ -87,22 +87,22 @@ cover_basic_blockst::cover_basic_blockst(const goto_programt &_goto_program)
 }
 
-unsigned cover_basic_blockst::block_of(goto_programt::const_targett t) const
+std::size_t cover_basic_blockst::block_of(goto_programt::const_targett t) const
```
Why all these changes to a `std::size_t` when the map itself (`block_mapt`) still maps to `unsigned`? And the commit message hasn't got a single word about this?!

romainbrenguier replied:
This was to fit the function type of the interface. I guess I will add a new commit at the beginning to do these type changes.

tautschnig commented on src/goto-instrument/cover_instrument_location.cpp:
```diff
@@ -17,12 +17,12 @@ Author: Daniel Kroening
     i_it->make_skip();
 
-  const unsigned block_nr = basic_blocks.block_of(i_it);
+  const std::size_t block_nr = basic_blocks.block_of(i_it);
```
Wrong commit?

romainbrenguier replied:
I added an initial commit for these changes

tautschnig commented on src/goto-instrument/cover_basic_blocks.cpp:
```diff
@@ -228,3 +160,46 @@ void cover_basic_blockst::update_covered_lines(block_infot &block_info)
+cover_basic_blocks_javat::instruction_of(const std::size_t block_nr) const
+{
+  INVARIANT(block_nr < block_infos.size(), "block number out of range");
+  return block_infos.at(block_nr);
```
Using `at` together with the above `INVARIANT` is unnecessary. Use `block_infos[block_nr]` or drop the `INVARIANT`. The same is true below.

tautschnig commented on src/goto-instrument/cover_basic_blocks.h:
```diff
@@ -130,4 +114,33 @@ class cover_basic_blockst final : public cover_blocks_baset
+  // map block number to first instruction of the block
+  std::vector<goto_programt::const_targett> block_infos;
+  // map java indexes to block indexes
+  std::unordered_map<dstringt, std::size_t, dstring_hash> index_to_block;
```
Use `irep_idt` (and `irep_id_hash`), never use `dstringt` directly.

romainbrenguier replied:
@tautschnig Alright if you say so. But may I ask, irep_idt is just a typedef dstringt, is there a reason we keep the two separated?

tautschnig replied:
The macro `USE_DSTRING` will make an `irep_idt` be a `dstringt`, otherwise an `irep_idt` is a `std::string`. It's a different question whether disabling `USE_DSTRING` is actually beneficial.

romainbrenguier replied:
I just try commenting out the `#define USE_DSTRING` and cbmc does not compile. So I guess it has not been tested for some time.

tautschnig replied:
Would you mind creating an issue, possibly including your build log? Certainly it hasn't been tested in a long time, and it should be properly evaluated whether it makes sense at all anymore. But there doesn't really seem any urgency to do this.

thk123 replied:
I've always hoped this would be fixed as IDEs have much better support for std::string - I thought I had raised an issue but that seems to be all in my head

romainbrenguier replied:
issue created here: https://github.com/diffblue/cbmc/issues/1837

tautschnig commented on src/goto-instrument/cover_basic_blocks.cpp:
```diff
@@ -228,3 +160,46 @@ void cover_basic_blockst::update_covered_lines(block_infot &block_info)
+    if(!index_to_block.count(bytecode_index))
+    {
+      block_infos.push_back(it);
+      index_to_block.emplace(bytecode_index, block_infos.size() - 1);
```
That's two lookups in quick succession. Instead:
````c++
auto &entry = index_to_block.insert({bytecode_index, block_infos.size() - 1});
if(entry.second)
  block_infos.push_back(it);
````

romainbrenguier replied:
Well spotted (except size-1 needs to be changed to size)

peterschrammel commented on src/goto-instrument/cover.cpp:
```diff
@@ -28,21 +28,27 @@ Date: May 2016
   const cover_instrumenterst &instrumenters,
-  message_handlert &message_handler)
+  message_handlert &message_handler,
+  const irep_idt mode)
```
const &

peterschrammel commented on src/goto-instrument/cover_instrument.h:
```diff
@@ -16,6 +16,7 @@ Author: Peter Schrammel
 
 #include <goto-programs/goto_model.h>
 #include <util/message.h>
+#include "cover_basic_blocks.h"
```
Changing the forward declaration below to cover_blocks_baset should be sufficient.

peterschrammel commented on src/goto-instrument/cover_basic_blocks.cpp:
```diff
@@ -228,3 +160,46 @@ void cover_basic_blockst::update_covered_lines(block_infot &block_info)
+    const auto &bytecode_index = it->source_location.get_java_bytecode_index();
+    if(!index_to_block.count(bytecode_index))
+    {
+      block_infos.push_back(it);
```
Where are the covered_lines set?

romainbrenguier replied:
They are not. Because I don't see how instructions with same bytecode index can correspond to different lines?

peterschrammel replied:
They must be set nonetheless in order not to change the output format.

romainbrenguier replied:
Oh right, I thought this was the same as get_line. This should be fixed in the last fixup commit https://github.com/diffblue/cbmc/pull/1830/commits/48fc65ceb2d5ec425e074e37a809237f4fa85a13

romainbrenguier committed:
Correct types from unsigned to std::size_t

The block numbers in cover_basic_blockst are obtained from
`block_info.size() - 1`, they should have type std::size_t rather than
unsigned.

Also mark const a couple of arguments that are constant.

romainbrenguier committed:
Use [] instead of `at` on vector

The index is checked against the size before end so there is no need to
use at there.

thk123 commented on src/goto-instrument/cover_basic_blocks.h:
```diff
@@ -18,44 +18,84 @@ Author: Daniel Kroening
 class message_handlert;
 
-class cover_basic_blockst final
+class cover_blocks_baset
```
Nit: move this into its own file

romainbrenguier replied:
I don't know, this is just an interface so very little code is added and this file is relatively small...

thk123 commented on src/goto-instrument/cover.cpp:
```diff
@@ -28,21 +28,27 @@ Date: May 2016
   const cover_instrumenterst &instrumenters,
-  message_handlert &message_handler)
+  message_handlert &message_handler,
+  const irep_idt mode)
```
The valid values for `mode` are `ID_java` or anything else? Consider an enum if you don't want to use a `bool enable_java_blocks`

peterschrammel replied:
`mode` is a standard concept used in the symbol table.

thk123 commented on src/goto-instrument/cover.cpp:
```diff
@@ -320,6 +327,7 @@ bool instrument_cover_goals(
 
   Forall_goto_functions(f_it, goto_functions)
   {
+    config->mode = symbol_table.lookup(f_it->first)->mode;
```
Eek local variable called `config` that's going to bite someone one day...

thk123 commented on src/goto-instrument/cover_basic_blocks.cpp:
```diff
@@ -191,12 +196,12 @@ cover_basic_blocks_javat::instruction_of(const std::size_t block_nr) const
 {
-  INVARIANT(block_nr < block_infos.size(), "block number out of range");
-  return block_infos[block_nr]->source_location;
+  INVARIANT(block_nr < block_locations.size(), "block number out of range");
```
Nit: probably a PRECONDITION

thk123 reviewed (commented):
It would be good to have some tests (in fact, for free this fixes some problems so tests introduced by #1812 plus something along the lines of the tests introduced in diffblue/test-gen#1474

romainbrenguier committed:
Adapt coverage test for new instrumentation

JBMC use to group together several lines within the same block.
This is no longer the case for the Java specific block instrumentation.
Each line is part of a different goal.

romainbrenguier commented:
I modified a test in https://github.com/diffblue/cbmc/pull/1830/commits/1083f74776d6d9039078ef656f219e988a876b62
which should show the difference between before and after these changes.
In https://github.com/diffblue/cbmc/pull/1812 I only see one unit test, and what is tested seems complicated compared to what the new instrumentation is doing. 
It difficult to have test showing the previous implementation would fail, because it depends on some non-predictable behaviour (like the order of some instructions which are added during instrumentation).
I will activate test similar to https://github.com/diffblue/test-gen/pull/1474 with the corresponding submodule update.

thk123 reviewed (approved):
@mgudemann do you agree that the unit test in #1812 is not relevant for this PR? 

@romainbrenguier In any case, it seemed like the virtual function resolution was consistently wrong (block would always go inside the block of the first if) - can this perhaps be used to have a regression/unit test?

mgudemann commented:
@thk123 I think I would make sense to adapt the test which would mainly mean removing the test on the added marker. But testing that there is no `ASSERT` after the initial guarded GOTO would still make sense, wouldn't it?
https://github.com/diffblue/cbmc/pull/1812/files#diff-0742ad545166f7e1ac81c0e158edb2b1R158

romainbrenguier commented:
@thk123 @mgudemann yes this makes sense, I will try adding the unit test

romainbrenguier committed:
Add unit tests for java block instrumentation

This checks that coverage instrumentation assertions are added at the
first instruction with given bytecode index.
And that there is no such assertion in the middle of instructions with
same bytecode index.

romainbrenguier commented:
@thk123 @mgudemann I added the unit test

thk123 commented on unit/java_bytecode/java_virtual_functions/virtual_functions.cpp:
```diff
@@ -115,6 +115,55 @@ SCENARIO(
+        REQUIRE(it != goto_program.instructions.end());
+
+        // Skip the first instruction which is a declaration with no location
+        ++it;
```
My feeling is you should do `std::next(goto_program.instructions.begin())` but then the require would be that the program includes at least two source locations, but currently you check the first one is not the end, then advance one and then dereference it, which might be a problem as it might be at the end. Using `std::next` when init'ing would prevent this.

thk123 commented on unit/java_bytecode/java_virtual_functions/virtual_functions.cpp:
```diff
@@ -115,6 +115,55 @@ SCENARIO(
+              // instruction with a new bytecode index begins with ASSERT
+              REQUIRE(it->type == goto_program_instruction_typet::ASSERT);
+              // the assertion corresponds to a line coverage goal
+              REQUIRE(!loc.get_basic_block_covered_lines().empty());
```
Prefer `REQUIRE_FALSE` when checking a negative

thk123 commented on unit/java_bytecode/java_virtual_functions/virtual_functions.cpp:
```diff
@@ -115,6 +115,55 @@ SCENARIO(
+          {
+            const source_locationt &loc = it->source_location;
+            REQUIRE(loc != source_locationt::nil());
+            REQUIRE(!loc.get_java_bytecode_index().empty());
```
Prefer `REQUIRE_FALSE` when checking a negative

thk123 reviewed (approved):
Unit test seems good - couple of suggestions but otherwise LGTM

peterschrammel reviewed (approved):
I think that the internal implementation could be further simplified and unified, but we can do that in a later PR when integrating https://github.com/diffblue/cbmc/pull/1439. Let's go ahead with that to solve first things first.

chrisr-diffblue commented on src/goto-instrument/cover.cpp:
```diff
@@ -28,21 +28,27 @@ Date: May 2016
   const cover_instrumenterst &instrumenters,
-  message_handlert &message_handler)
+  message_handlert &message_handler,
+  const irep_idt &mode)
```
Doxygen needs adding for the extra parameter, and as a minor nit, I might be tempted to put the new `mode` parameter before the `message_handler` parameter as the diff would be a fraction smaller, and it groups the parameters that have more meaning together i.e. leaving just the UI parameter tacked on the end.

chrisr-diffblue reviewed (approved):
One minor comment, but generally looks good so I'll approve and let you consider whether you agree with my comment or not.

romainbrenguier commented:
I have not merge this PR yet because I have seen a test failing on another repository after that change.  This other PR fixes the problem https://github.com/diffblue/cbmc/pull/1866 , so it would be better to wait for this second PR to be merged before merging this one.

romainbrenguier committed:
Update coverage goals in goto-diff test

This test mentions explicitely goal numbers, which are changed by the
java specific instrumentation.
This also fix potential problems with left braces that were not escaped
in the regular expression.
